src/mscorlib/src/System/Text/Encoding.cs

```diff
@@ -88,14 +88,54 @@ namespace System.Text
 #endif
     public abstract class Encoding : ICloneable
     {
-        private static volatile Encoding defaultEncoding;
-        private static volatile Encoding unicodeEncoding;
-        private static volatile Encoding bigEndianUnicode;
-        private static volatile Encoding utf7Encoding;
-        private static volatile Encoding utf8Encoding;
-        private static volatile Encoding utf32Encoding;
-        private static volatile Encoding asciiEncoding;
-        private static volatile Encoding latin1Encoding;
+        // We lazily-initialize the static Encoding properties by storing them in
+        // nested static classes. The static constructor, which holds the initialization
+        // logic, will only be run on the first access to a static member of the class
+        // (see dotnet/coreclr#1193 for more info). Thus, we get the benefits of both
+        // thread-safe and lock-free initialization.
+
+        // Please do not aggregate these classes into one static class. Otherwise,
+        // all of these encodings will be initialized when only one of them is needed.
+
+        private static class ASCIIHolder
+        {
```

Comment:
@jkotas I saw your suggestion earlier, however I did not follow it for reasons of code cleanliness (people might wonder why there is a static field being used nowhere in the file).

Perhaps what I'll do is make all the Encoding subclasses partial, and then have a file like `Encoding.StaticProperties.cs` (probably with a better name) and keep all of the static property-initialization logic in there. That way, the comment will not have to be repeated per-Encoding.